executor/src/witgen/affine_expression.rs

```diff
@@ -122,79 +122,146 @@ where
             },
         )
     }
 
     fn try_transfer_constraints(
         &self,
         known_constraints: &impl RangeConstraintSet<K, T>,
-    ) -> Option<EvalValue<K, T>> {
-        // We need the form X = a * Y + b * Z + ...
-        // where X is unconstrained and all others are bit-constrained.
-        let mut unconstrained = self
+    ) -> Option<(K, Constraint<T>)> {
+        // We are looking for X = a * Y + b * Z + ... or -X = a * Y + b * Z + ...
+        // where X is least constrained.
+
+        let (solve_for, solve_for_coefficient) = self
             .nonzero_coefficients()
-            .filter(|(i, _c)| known_constraints.range_constraint(*i).is_none());
-        let solve_for = unconstrained.next()?;
-        if unconstrained.next().is_some() {
-            return None;
-        }
-        if *solve_for.1 == 1.into() {
-            return (-self.clone()).try_transfer_constraints(known_constraints);
-        } else if *solve_for.1 != (-1).into() {
-            // We could try to divide by this in the future.
-            return None;
-        }
+            .filter(|(_i, c)| **c == (-1).into() || **c == 1.into())
+            .max_by_key(|(i, _c)| {
+                // Sort so that we get the least constrained variable.
+                known_constraints
+                    .range_constraint(*i)
+                    .map(|c| c.range_width())
+                    .unwrap_or_else(|| T::modulus())
+            })?;
 
-        // We can assume that nonzero coefficients is not empty, otherwise we could have solved
-        // the affine expression directly.
-        let parts = self
+        let summands = self
             .nonzero_coefficients()
-            .filter(|(i, _)| *i != solve_for.0)
+            .filter(|(i, _)| *i != solve_for)
             .map(|(i, coeff)| {
                 known_constraints
                     .range_constraint(i)
-                    .and_then(|con| con.multiple(*coeff))
-            });
-
-        parts
-            .reduce(|c1, c2| match (c1, c2) {
-                (Some(c1), Some(c2)) => c1.try_combine_sum(&c2),
-                _ => None,
-            })
-            .flatten()
-            .map(|con| {
-                EvalValue::incomplete_with_constraints(
-                    vec![(solve_for.0, Constraint::RangeConstraint(con))],
-                    IncompleteCause::NotConcrete,
-                )
+                    .map(|con| con.multiple(*coeff))
             })
+            .chain(
+                (self.offset != 0.into()).then_some(Some(RangeConstraint::from_value(self.offset))),
+            )
+            .collect::<Option<Vec<_>>>()?;
+        let mut constraint = summands.into_iter().reduce(|c1, c2| c1.combine_sum(&c2))?;
+        if *solve_for_coefficient == 1.into() {
+            constraint = -constraint;
+        }
+        if let Some(previous) = known_constraints.range_constraint(solve_for) {
+            if previous.clone().conjunction(&constraint) == previous {
```

Comment:
Maybe make `conjunction()` take reference instead of cloning? There isn't a particular strong case for it to take `self` instead of `&self`...